### PR TITLE
Add in retry-work to GPU OutOfCore Sort

### LINKED ISSUE #7934: [BUG] User app fails with OOM - GpuOutOfCoreSortIterator
**Describe the bug**
customer job is failing with the following OOM in GpuOutOfCoreSortIterator.  I worked around by increasing the shuffle partitions.

```
java.lang.OutOfMemoryError: Could not allocate native memory: std::bad_alloc: out_of_memory: RMM failure at:/home/jenkins/agent/workspace/jenkins-spark-rapids-jni_nightly-dev-346-cuda11/thirdparty/cudf/cpp/build/_deps/rmm-src/include/rmm/mr/device/arena_memory_resource.hpp:158: Maximum pool size exceeded
	at ai.rapids.cudf.Table.contiguousSplit(Native Method)
	at ai.rapids.cudf.Table.contiguousSplit(Table.java:2171)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.splitAfterSortAndSave(GpuSortExec.scala:358)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.$anonfun$mergeSortEnoughToOutput$7(GpuSortExec.scala:469)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.$anonfun$mergeSortEnoughToOutput$7$adapted(GpuSortExec.scala:468)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.withResource(GpuSortExec.scala:246)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.$anonfun$mergeSortEnoughToOutput$2(GpuSortExec.scala:468)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.$anonfun$mergeSortEnoughToOutput$2$adapted(GpuSortExec.scala:440)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.withResource(GpuSortExec.scala:246)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.mergeSortEnoughToOutput(GpuSortExec.scala:440)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.$anonfun$next$3(GpuSortExec.scala:520)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.withResource(GpuSortExec.scala:246)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.next(GpuSortExec.scala:519)
	at com.nvidia.spark.rapids.GpuOutOfCoreSortIterator.next(GpuSortExec.scala:246)
	at com.nvidia.spark.rapids.GpuKeyBatchingIterator.next(GpuKeyBatchingIterator.scala:157)
	at com.nvidia.spark.rapids.GpuHashAggregateIterator$$anon$2.next(aggregate.scala:476)
	at com.nvidia.spark.rapids.GpuHashAggregateIterator$$anon$2.next(aggregate.scala:468)
	at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$next$2(aggregate.scala:247)
	at scala.Option.getOrElse(Option.scala:189)
	at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:237)
	at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:182)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.partNextBatch(GpuShuffleExchangeExecBase.scala:318)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.hasNext(GpuShuffleExchangeExecBase.scala:340)
	at org.apache.spark.sql.rapids.RapidsShuffleThreadedWriterBase.$anonfun$write$2(RapidsShuffleInternalManagerBase.scala:281)
	at org.apache.spark.sql.rapids.RapidsShuffleThreadedWriterBase.$anonfun$write$2$adapted(RapidsShuffleInternalManagerBase.scala:274)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at org.apache.spark.sql.rapids.RapidsShuffleThreadedWriterBase.withResource(RapidsShuffleInternalManagerBase.scala:234)
	at org.apache.spark.sql.rapids.RapidsShuffleThreadedWriterBase.$anonfun$write$1(RapidsShuffleInternalManagerBase.scala:274)
	at org.apache.spark.sql.rapids.RapidsShuffleThreadedWriterBase.$anonfun$write$1$adapted(RapidsShuffleInternalManagerBase.scala:273)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at org.apache.spark.sql.rapids.RapidsShuffleThreadedWriterBase.withResource(RapidsShuffleInternalManagerBase.scala:234)
	at org.apache.spark.sql.rapids.RapidsShuffleThreadedWriterBase.write(RapidsShuffleInternalManagerBase.scala:273)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.$anonfun$runTask$3(ShuffleMapTask.scala:81)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.scheduler.ShuffleMapTask.$anonfun$runTask$1(ShuffleMapTask.scala:81)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:156)
	at org.apache.spark.scheduler.Task.$anonfun$run$1(Task.scala:125)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.scheduler.Task.run(Task.scala:95)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$13(Executor.scala:832)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1681)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:835)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at com.databricks.spark.util.ExecutorFrameProfiler$.record(ExecutorFrameProfiler.scala:110)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:690)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)

```

### PR DESCRIPTION
**Pull request by firestarman**
closes #7934

This PR adds in OOM retry to GpuOutOfCore iterator.

TODO

- [x] Add unit tests

### CODE REVIEW

**Review by revans2 — COMMENTED** [review]
I did a pass through this. It looks really good, but I didn't have time to really dig deeply into it. I had a few comments though.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala [hunk 1]
```diff
@@ -318,8 +335,7 @@ case class GpuOutOfCoreSortIterator(
         memUsed += ct.getBuffer.getLength
         val sp = SpillableColumnarBatch(ct, sorter.projectedBatchTypes,
           SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
-        sortedSize += sp.sizeInBytes
-        sorted.add(sp)
+        sortedCb = Some(sp)
       }
     } else {
       val hasFullySortedData = sortedOffset > 0
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:341]: `sortedTbl` is an argument, and will be closed after this `splitTableAfterSort` method returns.
You can find the closing in the `splitAfterSort` method.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala [hunk 2]
```diff
@@ -472,28 +533,36 @@ case class GpuOutOfCoreSortIterator(
...
+        sortedSize -= tmp.sizeInBytes
+        totalBytes += tmp.sizeInBytes
+        spillCbs += tmp
       }
       var memUsed = totalBytes
-      val ret = if (tables.length == 1) {
+      val ret = if (spillCbs.length == 1) {
         // We cannot concat a single table
-        sorter.removeProjectedColumns(tables.head)
+        withResource(spillCbs.head.getColumnarBatch()) { cb =>
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala]: done

**Comment by firestarman** [comment]
build

**Review by revans2 — APPROVED** [review]
This looks good. I would love to see some more tests.

**Comment by firestarman** [comment]
> This looks good. I would love to see some more tests.

Added in some unit tests. I will run the NDS next before merging this.

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
build

**Comment by firestarman** [comment]
Ran the NDS and no big perf regression is found.

**Pull request merged**